### PR TITLE
CDAP-6797: Adding authorization to programmatic API. Both programmatic and the REST APIs now go through the same DefaultSecureStoreService class, which does the authorization.

### PR DESCRIPTION
**Pull request by Nishith**

### CODE REVIEW

**Comment by Nishith** [comment]
Build under progress: http://builds.cask.co/browse/CDAP-RUT74-3

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/DefaultSecureStoreServiceTest.java [hunk 1]
```diff
@@ -140,38 +157,39 @@ public void testSecureStoreAccess() throws Exception {
...
+    secureKeyListEntries = secureStore.listSecureData(NamespaceId.DEFAULT.getNamespace());
     Assert.assertEquals(1, secureKeyListEntries.size());
 
     // Give BOB read access and verify that he can read the stored data
     SecurityRequestContext.setUserId(BOB.getName());
     grantAndAssertSuccess(NamespaceId.DEFAULT, BOB, EnumSet.of(Action.READ));
     grantAndAssertSuccess(secureKeyId1, BOB, EnumSet.of(Action.READ));
-    Assert.assertArrayEquals(VALUE1.getBytes(), secureStoreService.get(secureKeyId1).get());
-    secureKeyListEntries = secureStoreService.list(NamespaceId.DEFAULT);
+    Assert.assertEquals(VALUE1, new String(secureStore.getSecureData(NamespaceId.DEFAULT.getNamespace(), KEY1).get()));
```
Comment (bdmogal) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/DefaultSecureStoreServiceTest.java]: Bytes.toString

File: cdap-security/src/main/java/co/cask/cdap/security/store/DefaultSecureStoreService.java [hunk 2]
```diff
@@ -69,24 +72,21 @@
...
    * does not have access to the namespace or any of the keys in the namespace.
-   * @param namespaceId Id of the namespace we want the key list for.
    * @return A list of {@link SecureKeyListEntry} for all the keys visible to the user under the given namespace.
    * @throws NamespaceNotFoundException If the specified namespace does not exist.
    * @throws IOException If there was a problem reading from the store.
    *
    */
   @Override
-  public List<SecureKeyListEntry> list(NamespaceId namespaceId) throws Exception {
+  public final List<SecureStoreMetadata> listSecureData(String namespace) throws Exception {
```
Comment (bdmogal) [inline cdap-security/src/main/java/co/cask/cdap/security/store/DefaultSecureStoreService.java]: why change it to accept String?

Comment (bdmogal) [inline cdap-security/src/main/java/co/cask/cdap/security/store/DefaultSecureStoreService.java]: also, the return type changed? Is this an unrelated fix?

Comment (bdmogal) [inline cdap-security/src/main/java/co/cask/cdap/security/store/DefaultSecureStoreService.java]: oh it accepts string because it needs to be called from API?

File: cdap-security/src/main/java/co/cask/cdap/security/store/DefaultSecureStoreService.java [hunk 3]
```diff
@@ -95,68 +95,62 @@
...
    * @throws UnauthorizedException If the user does not have write permissions on the namespace.
    * @throws NamespaceNotFoundException If the specified namespace does not exist.
    * @throws AlreadyExistsException If the key already exists in the namespace. Updating is not supported.
    * @throws IOException If there was a problem storing the key to underlying provider.
    */
   @Override
-  public synchronized void put(SecureKeyId secureKeyId,
-                               SecureKeyCreateRequest secureKeyCreateRequest) throws Exception {
+  public final synchronized void putSecureData(String namespace, String name, String value, String description,
+                                         Map<String, String> properties) throws Exception {
```
Comment (bdmogal) [inline cdap-security/src/main/java/co/cask/cdap/security/store/DefaultSecureStoreService.java]: indentation

File: cdap-security/src/main/java/co/cask/cdap/security/store/DefaultSecureStoreService.java [hunk 4]
```diff
@@ -95,68 +95,62 @@
...
     if (Strings.isNullOrEmpty(value)) {
       throw new BadRequestException("The data field should not be empty. This is the data that will be stored " +
                                       "securely.");
     }
 
-    secureStoreManager.putSecureData(secureKeyId.getNamespace(), secureKeyId.getName(), value, description,
-                           secureKeyCreateRequest.getProperties());
-    authorizer.grant(secureKeyId, principal, ImmutableSet.of(Action.ALL));
+    secureStoreManager.putSecureData(namespace, name, value, description, properties);
+    privilegesManager.grant(new SecureKeyId(namespace, name), principal, ImmutableSet.of(Action.ALL));
```
Comment (bdmogal) [inline cdap-security/src/main/java/co/cask/cdap/security/store/DefaultSecureStoreService.java]: should grant first. See https://wiki.cask.co/display/CE/Authorization+-+CDAP+3.5#Authorization-CDAP3.5-AuthorizationPolicies. Also revoke all on the key as the first step

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/SecureStoreHandler.java [hunk 5]
```diff
@@ -86,7 +89,7 @@ public void delete(HttpRequest httpRequest, HttpResponder httpResponder, @PathPa
   public void get(HttpRequest httpRequest, HttpResponder httpResponder, @PathParam("namespace-id") String namespace,
                   @PathParam("key-name") String name) throws Exception {
     SecureKeyId secureKeyId = new SecureKeyId(namespace, name);
-    String data = new String(secureStoreService.get(secureKeyId).get(), StandardCharsets.UTF_8);
+    String data = new String(secureStore.getSecureData(namespace, name).get(), StandardCharsets.UTF_8);
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/SecureStoreHandler.java]: Strictly speaking, you don't need to convert the `byte[]` to `String` and call the `sendString` (which in turn will turn the `String` back to `byte[]`). You can simply do `responder.sendContent(OK, ChannelBuffers.wrappedBuffer(data), "text/plain; charset=utf-8", null);`

File: cdap-security/src/main/java/co/cask/cdap/security/guice/SecureStoreModules.java [hunk 6]
```diff
@@ -59,14 +58,22 @@
...
+  public static final String DELEGATE_SECURE_STORE_MANAGER = "delegateSecureStoreManager";
 
   @Override
   public final Module getInMemoryModules() {
     return new AbstractModule() {
       @Override
       protected void configure() {
-        bind(SecureStore.class).toProvider(new TypeLiteral<StoreProvider<SecureStore>>() { });
-        bind(SecureStoreManager.class).toProvider(new TypeLiteral<StoreProvider<SecureStoreManager>>() { });
+        bind(SecureStore.class)
```
Comment (chtyim) [inline cdap-security/src/main/java/co/cask/cdap/security/guice/SecureStoreModules.java:70]: Should use `PrivateModule` instead and have these two `DELEGATE` binding as private binding and only exposes the unannotated `SecureStore` and `SecureStoreManager`.

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/DefaultSecureStoreServiceTest.java (hunk 1 above)
Comment (Nishith) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/DefaultSecureStoreServiceTest.java]: Any particular reason?

Comment (bdmogal) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/DefaultSecureStoreServiceTest.java]: encoding

File: cdap-security/src/main/java/co/cask/cdap/security/store/DefaultSecureStoreService.java (hunk 2 above)
Comment (Nishith) [inline cdap-security/src/main/java/co/cask/cdap/security/store/DefaultSecureStoreService.java]: Yes, the API takes a String.

**Comment by Nishith** [comment]
Green build: http://builds.cask.co/browse/CDAP-RUT74-13

**Comment by Nishith** [comment]
Mini-ITN: http://builds.cask.co/browse/CDAP-ITM0-21

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/DefaultSecureStoreServiceTest.java [hunk 7]
```diff
@@ -100,6 +111,11 @@ public Boolean call() throws Exception {
     authorizer.revoke(NamespaceId.DEFAULT, ALICE, Collections.singleton(Action.READ));
   }
 
+  private static void waitForService(String service) {
+    EndpointStrategy endpointStrategy = new RandomEndpointStrategy(discoveryServiceClient.discover(service));
+    Preconditions.checkNotNull(endpointStrategy.pick(5, TimeUnit.SECONDS),
+                               "%s service is not up after 5 seconds", service);
+  }
```
Comment (bdmogal) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/DefaultSecureStoreServiceTest.java:116]: nit: add newline after method

File: cdap-app-fabric/src/test/java/co/cask/cdap/security/SecureStoreTest.java [hunk 8]
```diff
@@ -22,23 +22,20 @@
...
 import java.lang.reflect.Type;
-import java.util.ArrayList;
-import java.util.List;
+import java.util.HashMap;
 import java.util.Map;
 
 public class SecureStoreTest extends AppFabricTestBase {
   private static final Gson GSON = new Gson();
-  private static Type listType = new TypeToken<ArrayList<SecureKeyListEntry>>() { }.getType();
+  private static Type mapType = new TypeToken<Map<String, String>>() { }.getType();
```
Comment (bdmogal) [inline cdap-app-fabric/src/test/java/co/cask/cdap/security/SecureStoreTest.java]: final. Also, ALL_CAPS

File: cdap-kms/src/main/java/co/cask/cdap/security/store/KMSSecureStore.java [hunk 9]
```diff
@@ -181,13 +182,11 @@ public void deleteSecureData(String namespace, String name) throws Exception {
     if (metadatas.length != keysInNamespace.size()) {
       throw new ConcurrentModificationException("A key was deleted while listing was in progress. Please try again.");
     }
-    List<SecureStoreMetadata> secureStoreMetadatas = new ArrayList<>(metadatas.length);
+//    List<SecureStoreMetadata> secureStoreMetadatas = new ArrayList<>(metadatas.length);
```
Comment (bdmogal) [inline cdap-kms/src/main/java/co/cask/cdap/security/store/KMSSecureStore.java]: remove?

File: cdap-security/src/main/java/co/cask/cdap/security/store/DefaultSecureStoreService.java [hunk 10]
```diff
@@ -69,24 +69,20 @@
   /**
    * Lists all the secure keys in the given namespace that the user has access to. Returns an empty list if the user
    * does not have access to the namespace or any of the keys in the namespace.
-   * @param namespaceId Id of the namespace we want the key list for.
-   * @return A list of {@link SecureKeyListEntry} for all the keys visible to the user under the given namespace.
+   * @return map
```
Comment (bdmogal) [inline cdap-security/src/main/java/co/cask/cdap/security/store/DefaultSecureStoreService.java]: explain?

File: cdap-security/src/main/java/co/cask/cdap/security/store/DefaultSecureStoreService.java [hunk 11]
```diff
@@ -95,68 +91,62 @@
...
    * @throws NotFoundException If the key is not found in the store.
    * @throws IOException If there was a problem reading from the store.
    * @throws UnauthorizedException If the user does not have READ permissions on the secure key.
    */
   @Override
-  public SecureStoreData get(SecureKeyId secureKeyId) throws Exception {
+  public final SecureStoreData getSecureData(String namespace, String name) throws Exception {
     Principal principal = authenticationContext.getPrincipal();
     final Predicate<EntityId> filter;
     filter = authorizationEnforcer.createFilter(principal);
```
Comment (bdmogal) [inline cdap-security/src/main/java/co/cask/cdap/security/store/DefaultSecureStoreService.java]: nit: combine with previous line

File: cdap-security/src/main/java/co/cask/cdap/security/store/DefaultSecureStoreService.java [hunk 12]
```diff
@@ -95,68 +91,62 @@
...
    */
   @Override
-  public void delete(SecureKeyId secureKeyId) throws Exception {
+  public final void deleteSecureData(String namespace, String name) throws Exception {
     Principal principal = authenticationContext.getPrincipal();
+    SecureKeyId secureKeyId = new SecureKeyId(namespace, name);
     authorizationEnforcer.enforce(secureKeyId, principal, Action.ADMIN);
-    secureStoreManager.deleteSecureData(secureKeyId.getNamespace(), secureKeyId.getName());
-    authorizer.revoke(secureKeyId);
+    privilegesManager.revoke(secureKeyId);
```
Comment (bdmogal) [inline cdap-security/src/main/java/co/cask/cdap/security/store/DefaultSecureStoreService.java]: revoke should happen after deletion is successful, so there are no orphan secure keys (in case deletion fails) to which no one has access.

**Comment by bdmogal** [comment]
@Nishith some minor comments. However, what was the reason for changing the list api to return a map?

**Comment by Nishith** [comment]
The design had always been to return name and description only in list. For REST it was handled in the handler but it was incorrect for the programmatic API.

**Comment by bdmogal** [comment]
LGTM
